### DISCUSSION TITLE
How do I create rearrangeable tabs in the title bar (like Firefox)?

### DISCUSSION THREAD

**Discussion by alshdavid** [gtk-rs/gtk4-rs#1378] [Q&A]
![image](https://user-images.githubusercontent.com/12656294/232945374-a14d4b32-a068-4d73-9bea-02a0e8db49a3.png)

I am new to gtk and gkt-rs so sorry for the basic question. I am trying to create an application that has tabs in the title bar that mimic the behaviour seen in Firefox's tabs. That is;

- One process can have multiple window instances
- Can rearrange tabs on the same window instance
- Can drag tab off a window instance to produce a new window instance
- Can drag tab from one windows instance onto another window instance
- Middle click closes a tab

Looking through the [Widget Gallery](https://docs.gtk.org/gtk4/visual_index.html#windows), I am not seeing any supported method of implementing this.

Is this something you have to implement by hand? If so, how does one do it such that it's in line with the adwaita best practices?

**Comment by bilelmoussaoui** [comment]
Use https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/struct.TabBar.html
